codex/codex.nim

```diff
@@ -141,14 +141,17 @@ proc stop*(s: CodexServer) {.async.} =
 
   s.runHandle.complete()
 
-proc new*(T: type CodexServer, config: CodexConf, privateKey: CodexPrivateKey): T =
+proc createCodexServer(
```

Comment:
@emizzle I renamed it and made it private. What do you think?

Reply:
> Ah, actually this one isn't directly related to the nimsuggest fix.
>
>I used it for tracking down the issue as the new proc's are a generic overload and nimsuggest won't fully type check them until they're used somewhere.
>
>For that reason with large complicated it's helpful to split into a concrete proc with a new wrapper.
>
>I could revert it back to new, but I think making the newCodexServer private would make new the external api but provide the compile checking.

`CodexServer.new` is used on https://github.com/codex-storage/nim-codex/blob/master/codex.nim#L92. Shouldn't that allow for nimsuggest to compile check the `new` proc without needing the non-exported `newCodexServer`?

Whatever ends up being settled on, we should make an effort to be consistent across the codebase (change all instances), and document it somewhere.

> @emizzle I renamed it and made it private. What do you think?

My personal opinion is to have neither the `newCodexServer` nor `createCodexServer` workarounds and keep it as it is :) My suspicion is that nimsuggest shouldn't have issues compile checking the way it is, and that possibly there's a nimsuggest fix that could be looked into. But I don't know enough about nimsuggest to comment beyond speculation.

Reply:
> CodexServer.new is used on https://github.com/codex-storage/nim-codex/blob/master/codex.nim#L92. Shouldn't that allow for nimsuggest to compile check the new proc without needing the non-exported newCodexServer?

Yes it will compile & check it from that file. However, with the normal nimsuggest setup if you edit and modify the codex/codex.nim file it won't catch errors there. In this case it helped hide some other issues. 

> My personal opinion is to have neither the newCodexServer nor createCodexServer workarounds and keep it as it is :) 

I can see that point. Skimming through I see that most (but not all) of the `new` constructors are simple constructors. 

My inclination would be to have a separate name for more complicated constructors like this. A proc of `createCodexServer` or similar would help imply it's doing something more than a normal `new` proc. Though that could be left over habit from older style C++ constructors.

> Whatever ends up being settled on, we should make an effort to be consistent across the codebase (change all instances), and document it somewhere.

That's a good point. 

> My suspicion is that nimsuggest shouldn't have issues compile checking the way it is, and that possibly there's a nimsuggest fix that could be looked into. But I don't know enough about nimsuggest to comment beyond speculation.

Yah, I think it'd be a useful feature for nimsuggest (really nim). 

I believe the reason it's not type checked at declaration is that, is that actually a generic proc of type `new*[T: CodexServer](...): T` under the hood. 

Though in this instance the compiler has enough type information to be resolved. Given the ubiquity of the `new` pattern I'd think it'd make sense to get it added to nimsuggest / nim.